service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/EncounterExtractor.java

```diff
@@ -0,0 +1,52 @@
+package uk.nhs.adaptors.gp2gp.ehr.utils;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.hl7.fhir.dstu3.model.Bundle;
+import org.hl7.fhir.dstu3.model.Encounter;
+import org.hl7.fhir.dstu3.model.ListResource;
+import org.hl7.fhir.dstu3.model.Reference;
+import org.hl7.fhir.dstu3.model.ResourceType;
+
+public final class EncounterExtractor {
+    private static final String CONSULTATION_LIST_CODE = "1149501000000101";
+
+    private EncounterExtractor() {
+    }
+
+    public static List<String> extractEncounterReferences(List<Bundle.BundleEntryComponent> entries) {
+        return entries.stream()
+            .filter(entry -> entry.getResource().getResourceType().equals(ResourceType.List))
+            .map(entry -> (ListResource) entry.getResource())
+            .filter(EncounterExtractor::isConsultationList)
+            .findFirst()
+            .map(listResource -> extractEncounterReferencesFromListResource(listResource.getEntry()))
+            .orElse(Collections.emptyList());
+    }
+
+    public static List<Encounter> extractEncountersByReferences(List<Bundle.BundleEntryComponent> entries,
+            List<String> encounterReferences) {
+        return entries.stream()
+            .filter(entry -> entry.getResource().getResourceType().equals(ResourceType.Encounter))
+            .map(entry -> (Encounter) entry.getResource())
+            .filter(encounter -> encounterReferences.contains(encounter.getId()))
+            .collect(Collectors.toList());
+    }
+
+    private static boolean isConsultationList(ListResource listResource) {
+        return listResource.getCode()
+            .getCoding()
+            .get(0)
```

Comment:
Are we sure about single code only? I would check if any of the codes match the consultation code.

Reply:
Also protect against the missing code/coding cases